File: src/utils/imageUploader.ts
import axios from "axios";
import * as fs from "fs";
import * as path from "path";

export class ImageUploader {
  private apiToken: string;

  constructor(apiToken: string) {
    this.apiToken = apiToken;
  }

  /**
   * Upload an image to Printify
   */
  async uploadImage(imagePath: string): Promise<{ id: string; url: string; preview_url: string }> {
    try {
      console.log(`📤 Uploading image: ${imagePath}`);

      // Check if file exists
      if (!fs.existsSync(imagePath)) {
        throw new Error(`Image file not found: ${imagePath}`);
      }

      // Read the image file
      const imageBuffer = fs.readFileSync(imagePath);
      const fileName = path.basename(imagePath);

      // Create form data
      const FormData = require("form-data");
      const form = new FormData();
      form.append("file", imageBuffer, {
        filename: fileName,
        contentType: this.getContentType(fileName),
      });

      // Upload to Printify
      const response = await axios.post("https://api.printify.com/v1/uploads/images.json", form, {
        headers: {
          Authorization: `Bearer ${this.apiToken}`,
          "User-Agent": "EdenPrintify/1.0.0",
          ...form.getHeaders(),
        },
      });

      const uploadData = response.data;
      console.log(`✅ Image uploaded successfully: ${uploadData.id}`);

      return {
        id: uploadData.id,
        url: uploadData.url,
        preview_url: uploadData.preview_url,
      };
    } catch (error) {
      console.error(`❌ Failed to upload image: ${error}`);
      throw error;
    }
  }

  /**
   * Create a simple test image if none exists
   */
  async createTestImage(): Promise<string> {
    const testImagePath = "./test-image.png";

    if (fs.existsSync(testImagePath)) {
      console.log(`✅ Test image already exists: ${testImagePath}`);
      return testImagePath;
    }

    console.log(`🎨 Creating test image: ${testImagePath}`);

    // Create a simple PNG image using a data URL
    const { createCanvas } = require("canvas");
    const canvas = createCanvas(500, 500);
    const ctx = canvas.getContext("2d");

    // Draw a simple design
    ctx.fillStyle = "#FF6B6B";
    ctx.fillRect(0, 0, 500, 500);

    ctx.fillStyle = "#FFFFFF";
    ctx.font = "48px Arial";
    ctx.textAlign = "center";
    ctx.fillText("Test Design", 250, 250);

    // Save the image
    const buffer = canvas.toBuffer("image/png");
    fs.writeFileSync(testImagePath, buffer);

    console.log(`✅ Test image created: ${testImagePath}`);
    return testImagePath;
  }

  /**
   * Get content type based on file extension
   */
  private getContentType(fileName: string): string {
    const ext = path.extname(fileName).toLowerCase();
    const contentTypes: { [key: string]: string } = {
      ".png": "image/png",
      ".jpg": "image/jpeg",
      ".jpeg": "image/jpeg",
      ".gif": "image/gif",
      ".webp": "image/webp",
    };
    return contentTypes[ext] || "image/png";
  }

  /**
   * Upload multiple images and return their data
   */
  async uploadImages(imagePaths: string[]): Promise<Array<{ id: string; url: string; preview_url: string }>> {
    const results = [];
    for (const imagePath of imagePaths) {
      try {
        const result = await this.uploadImage(imagePath);
        results.push(result);
      } catch (error) {
        console.error(`❌ Failed to upload ${imagePath}:`, error);
        throw error;
      }
    }
    return results;
  }
}

File: python/src/utils/image_uploader.py
"""
Image Uploader
Handles image uploads to Printify
"""

import os
import requests
import tempfile
from typing import Dict, Any
from PIL import Image, ImageDraw, ImageFont


class ImageUploader:
    """Handles image uploads to Printify"""
    
    def __init__(self, api_token: str):
        self.api_token = api_token
        self.base_url = "https://api.printify.com/v1"
        self.headers = {
            "Authorization": f"Bearer {api_token}",
            "User-Agent": "EdenPrintify/1.0.0"
        }
    
    def upload_image(self, image_path: str) -> Dict[str, str]:
        """Upload an image to Printify"""
        try:
            # Check if file exists
            if not os.path.exists(image_path):
                raise FileNotFoundError(f"Image file not found: {image_path}")
            
            # Prepare the upload data
            with open(image_path, 'rb') as f:
                files = {'file': (os.path.basename(image_path), f, 'image/png')}
                
                response = requests.post(
                    f"{self.base_url}/uploads/images.json",
                    headers=self.headers,
                    files=files,
                    timeout=60
                )
            
            response.raise_for_status()
            upload_data = response.json()
            
            return {
                'id': upload_data['id'],
                'url': upload_data.get('url', upload_data.get('preview_url')),
                'preview_url': upload_data['preview_url']
            }
            
        except Exception as e:
            print(f"Failed to upload image {image_path}: {e}")
            raise
    
    def create_test_image(self) -> str:
        """Create a simple test image for testing uploads"""
        try:
            # Create a simple test image
            width, height = 400, 400
            image = Image.new('RGB', (width, height), color='white')
            draw = ImageDraw.Draw(image)
            
            # Add some text
            try:
                # Try to use a default font
                font = ImageFont.load_default()
            except:
                font = None
            
            text = "Test Image\nEden Printify"
            bbox = draw.textbbox((0, 0), text, font=font) if font else (0, 0, 200, 50)
            text_width = bbox[2] - bbox[0]
            text_height = bbox[3] - bbox[1]
            
            x = (width - text_width) // 2
            y = (height - text_height) // 2
            
            draw.text((x, y), text, fill='black', font=font)
            
            # Add a border
            draw.rectangle([0, 0, width-1, height-1], outline='black', width=2)
            
            # Save to temporary file
            temp_file = tempfile.NamedTemporaryFile(suffix='.png', delete=False)
            image.save(temp_file.name, 'PNG')
            temp_file.close()
            
            return temp_file.name
            
        except Exception as e:
            print(f"Failed to create test image: {e}")
            raise 